include/net/xia_ether.h

```diff
@@ -240,6 +243,105 @@ static int attach_neigh_addr_to_fib_entry(struct fib_xid_ether_main
 	return 0;
 }
 
+/* Template struct containing hdr manipulation funcs. */
+struct xia_header_ops {
+	/* Creates a new L2 header. */
+	int (*create) (struct sk_buff *skb, struct net_device *dev,
+			   unsigned short type, const void *daddr,
+			   const void *saddr, unsigned int len);
+
+	/* Extracts hrdw address from pkt. */
+	int (*parse)(const struct sk_buff *skb, unsigned char *haddr);
+
+	/* Creates a cached hdr and alocates to a mfxid. */
+	int (*cache)(const struct fib_xid_ether_main *mfxid, struct hh_cache *hh, __be16 type);
+
+	/* Updates the cached hdr on change of address
+	 * type : 1 -> destionation addr
+	 *        0 -> source addr
+	 */
+	void (*cache_update)(struct hh_cache *hh, const struct net_device *dev,
+				const unsigned char *haddr, const int type);
+};
+
+int xia_ether_header_cache(const struct fib_xid_ether_main *mfxid, 
+        struct hh_cache *hh, __be16 type)
+{
+	struct ethhdr *eth;
+	const struct net_device *dev = mfxid->host_interface;
+
+	eth = (struct ethhdr *)
+	    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+
+	if (type != htons(ETH_P_XIP))
+		return -1;
+
+	eth->h_proto = type;
+	memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+	memcpy(eth->h_dest, mfxid->neigh_addr->ha, ETH_ALEN);
+	hh->hh_len = ETH_HLEN;
+	return 0;
+}
+
+void xia_ether_header_cache_update(struct hh_cache *hh,
+			     const struct net_device *dev,
+			     const unsigned char *haddr,
+			     const int type)
+{
+	struct ethhdr *eth;
+	eth = (struct ethhdr *)
+		    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+	if(type)
+		memcpy(eth->h_dest , haddr, ETH_ALEN);
+	else
+		memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+}
+
+const struct xia_header_ops xia_ether_hdr_ops ____cacheline_aligned = {
+	.create         = eth_header,
+	.parse          = eth_header_parse,
+	.cache          = xia_ether_header_cache,
+	.cache_update   = xia_ether_header_cache_update,
+};
+
+/* Cached Hdr functions. */
+static void mfxid_hh_init(struct fib_xid_ether_main *mfxid)
+{
+	__be16 prot = htons(ETH_P_XIP);
+	struct hh_cache	*hh = &mfxid->cached_hdr;
+
+	write_lock_bh(&mfxid->chdr_lock);
```

Comment:
The purpose of providing chdr_lock was that even if neighbour is being accessed by multiple processes, the ability to change(not only initialize) the cached header lock is at once only with one writer. Till now according to my needs we need not care about the different writers changing the cached headers but for nwp i think it might be a different case(like when the ACK packets about the L2 address change arrive, then the last writer's cached header will be the final cached header which might be wrong or some other case).
My only point being that this field could be deprecated during the NWP development or it might come in handy at that point of time. I will remove this entry if you still don't find any use in keeping it there.